internal/validators/rbac.go

```diff
@@ -0,0 +1,145 @@
+// If we keep this approach, this will become just the interface, not 2.
+type roleAssignmentAPI2 interface {
+	ListRoleAssignmentsForScope(scope string, filter *string) ([]*armauthorization.RoleAssignment, error)
+}
+
+type RBACRuleService struct {
+	log              logr.Logger
+	api              roleAssignmentAPI2
+	getRoleLookupMap roleLookupMapProvider
+}
+
+func NewRBACRuleService(log logr.Logger, api roleAssignmentAPI2, roleLookupMapProvider roleLookupMapProvider) *RBACRuleService {
+	return &RBACRuleService{
+		log:              log,
+		api:              api,
+		getRoleLookupMap: roleLookupMapProvider,
+	}
+}
+
+// ReconcileRBACRule reconciles a role assignment rule from a validation config.
+func (s *RBACRuleService) ReconcileRBACRule(rule v1alpha1.RBACRule) (*vapitypes.ValidationResult, error) {
+
+	// Build the default ValidationResult for this role assignment rule.
+	state := vapi.ValidationSucceeded
+	latestCondition := vapi.DefaultValidationCondition()
+	latestCondition.Message = "Security principal has all required roles."
+	latestCondition.ValidationRule = fmt.Sprintf("%s-%s", vapiconstants.ValidationRulePrefix, rule.SecurityPrincipalID)
+	latestCondition.ValidationType = constants.ValidationTypeRBAC
+	validationResult := &vapitypes.ValidationResult{Condition: &latestCondition, State: &state}
+
+	failures := make([]string, 0)
+
+	for i, set := range rule.Permissions {
+		s.log.V(0).Info("Processing permission set of rule.", "set #", i+1)
+		if err := s.processPermissionSet(set, rule.SecurityPrincipalID, &failures); err != nil {
+			// Code this is returning to will take care of changing the validation result to a
+			// failed validation, using the error returned.
+			return validationResult, err
+		}
+	}
+
+	if len(failures) > 0 {
+		state = vapi.ValidationFailed
+		latestCondition.Failures = failures
+		latestCondition.Message = "Security principal missing one or more required roles."
+		latestCondition.Status = corev1.ConditionFalse
+	}
+
+	return validationResult, nil
+}
+
+// processPermissionSet processes a permission set from the rule.
+//   - set: The set to process.
+//   - principalID: The ID of the security principal to use in the filter. This comes from the rule
+//     the set is part of.
+//   - failures: The list of failures being built up while processing the entire rule. Must be
+//     non-nil.
+func (s *RBACRuleService) processPermissionSet(set v1alpha1.PermissionSet, principalID string, failures *[]string) error {
+
+	foundRoleNames := make(map[string]bool)
+
+	// Get all role assignments that apply to the specified scope where the member of the role
+	// assignment is the specified security principal. In this query, "principalId" must be a UUID,
+	// so this shouldn't have any injection vulnerabilities.
+	//
+	// Note that this also returns role assignments that assign the role because the scope is a
+	// surrounding scope (e.g. the subscription the scope is contained within), not just the scope
+	// itself.
```

Comment:
What do you think of this? It means that for a rule where the permission set defines a particular resource group as the scope, if the role had been assigned to the security principal for that scope indirectly because it was assigned to the security principal for the scope of the subscription that resource group was a part of, this would *pass*. Is that bad? Passed when it should fail? If so, I think we can tighten this up by using a richer filter here. I think they support adding scope to the filter too (not just as the argument below).

Reply:
I think the current behaviour is fine. We are interested in ensuring that User A has permissions X,Y,Z on resource group B. I don't think we need to concern ourselves with the scope of the role assignment that granted those permissions.